src/gt4py/next/iterator/runtime.py

```diff
@@ -49,75 +50,81 @@ class UnstructuredDomain(dict): ...
 fendef_embedded: Optional[Callable[[types.FunctionType], None]] = None
 
 
+@dataclasses.dataclass
 class FendefDispatcher:
-    def __init__(self, function: types.FunctionType, executor_kwargs: dict):
-        self.function = function
-        self.executor_kwargs = executor_kwargs
+    definition: types.FunctionType
+    offset_provider: Optional[common.OffsetProvider]
+    column_axis: Optional[common.Dimension]
 
-    def itir(
-        self,
-        *args,
-        fendef_codegen: Optional[Callable[[types.FunctionType], FencilDefinition]] = None,
-        debug=False,
-        **kwargs,
-    ):
-        args, kwargs = self._rewrite_args(args, kwargs)
-
-        # For consistency with `__call__` we also allow these keyword arguments, but ignore them
-        # here, as they are not used for code generation.
-        for ignored_kwarg in ["offset_provider", "lift_mode", "column_axis"]:
-            kwargs.pop(ignored_kwarg, None)
-
-        if fendef_codegen is None:
-            # TODO(ricoh): refactor so that `tracing` does not import this module
-            #   and can be imported top level. Then set `fendef_tracing` as a
-            #   proper default value, instead of using `None` as a sentinel.
-            from gt4py.next.iterator.tracing import trace_fencil_definition
-
-            fendef_codegen = trace_fencil_definition
-        fencil_definition = fendef_codegen(self.function, args, **kwargs)
-        if debug:
+    def itir(self, *args):
+        # TODO(ricoh): refactor so that `tracing` does not import this module
+        #   and can be imported top level. Then set `fendef_tracing` as a
+        #   proper default value, instead of using `None` as a sentinel.
+        from gt4py.next.iterator.tracing import trace_fencil_definition
+
+        fencil_definition = trace_fencil_definition(self.definition, args)
+
+        if config.DEBUG:
             devtools.debug(fencil_definition)
         return fencil_definition
 
-    def __call__(self, *args, backend: Optional[ProgramExecutor] = None, **kwargs):
-        args, kwargs = self._rewrite_args(args, kwargs)
+    def __call__(
+        self,
+        *args,
+        backend: Optional[next_backend.Backend] = None,
+        offset_provider=None,
+        column_axis=None,
+    ):
+        offset_provider = offset_provider or self.offset_provider
+        column_axis = column_axis or self.column_axis
 
         if backend is not None:
-            ensure_processor_kind(backend, ProgramExecutor)
-            backend(self.itir(*args, **kwargs), *args, **kwargs)
+            itir_node = self.itir(*args)
+
+            # TODO(tehrengruber): remove cirular dependency and place import at the top of the file
+            from gt4py.next import backend as next_backend
```

Comment:
where is it coming from?

Reply:
`runtime` -> `backend` -> `ffront` passes -> `ffront.fbuiltins` -> `runtime.Offset` used in `FieldOffset`
`FieldOffset` already has an open that touches on this.